File: src/main/java/leetcode/lists/topinterview/LeetCode230.java
package leetcode.lists.topinterview;

import common.TreeNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/28 22:37
 */
public class LeetCode230 {

    public int kthSmallest(TreeNode root, int k) {
        if (root == null) {
            return -1;
        }
        Stack<TreeNode> stack = new Stack<>();
        while (root != null || !stack.empty()) {
            while (root != null) {
                stack.add(root);
                root = root.left;
            }
            root = stack.pop();
            k--;
            if (k == 0) {
                break;
            }
            root = root.right;
        }
        return root.val;
    }
}

File: src/main/go/leetcode/lists/topinterview/LeetCode230.go
package topinterview

func kthSmallest(root *TreeNode, k int) int {

	if root == nil {
		return 0
	}
	cnt := 0
	res := 1
	var dfs func(root *TreeNode)
	dfs = func(root *TreeNode) {
		if root == nil {
			return
		}
		dfs(root.Left)
		cnt++
		if cnt == k {
			res = root.Val
			return
		}
		dfs(root.Right)
	}
	dfs(root)
	return res
}
